src/VisualStudio/Core/Def/Implementation/Preview/PreviewUpdater.cs

```diff
@@ -52,34 +50,65 @@ public void UpdateView(Document document, SpanChange spanSource)
-                var updatedSolution = _previewWorkspace.CurrentSolution.WithDocumentText(document.Id, document.GetTextAsync().Result);
-                var updatedDocument = updatedSolution.GetDocument(document.Id);
-
-                ApplyDocumentToBuffer(updatedDocument, spanSource, out container, out documentBackedByTextBuffer);
+                // Close the current document in preview solution.
+                if (_isCurrentAdditionalDocument)
+                {
+                    _previewWorkspace.CloseAdditionalDocument(_currentDocument, _previewWorkspace.CurrentSolution.GetAdditionalDocument(_currentDocument).GetTextAsync().Result);
+                }
+                else
+                {
+                    _previewWorkspace.CloseDocument(_currentDocument, _previewWorkspace.CurrentSolution.GetDocument(_currentDocument).GetTextAsync().Result);
+                }
+
+                // Put the new document into the current preview solution.
+                TextDocument updatedDocument;
+                if (isAdditionalDocument)
+                {
+                    var updatedSolution = _previewWorkspace.CurrentSolution.WithAdditionalDocumentText(document.Id, document.GetTextAsync().Result);
+                    updatedDocument = updatedSolution.GetAdditionalDocument(document.Id);
+                }
+                else
+                {
+                    var updatedSolution = _previewWorkspace.CurrentSolution.WithDocumentText(document.Id, document.GetTextAsync().Result);
+                    updatedDocument = updatedSolution.GetDocument(document.Id);
+                }
+
+                ApplyDocumentToBuffer(updatedDocument, spanSource, isAdditionalDocument, out container, out documentBackedByTextBuffer);
 
                 _previewWorkspace.TryApplyChanges(documentBackedByTextBuffer.Project.Solution);
-                _previewWorkspace.OpenDocument(document.Id);
+                OpenDocument(document.Id, isAdditionalDocument);
                 _currentDocument = document.Id;
             }
             else
             {
                 _currentDocument = document.Id;
 
-                ApplyDocumentToBuffer(document, spanSource, out container, out documentBackedByTextBuffer);
+                ApplyDocumentToBuffer(document, spanSource, isAdditionalDocument, out container, out documentBackedByTextBuffer);
 
                 _previewWorkspace = new PreviewDialogWorkspace(documentBackedByTextBuffer.Project.Solution);
-                _previewWorkspace.OpenDocument(document.Id);
+                OpenDocument(document.Id, isAdditionalDocument);
+            }
+
+            _isCurrentAdditionalDocument = isAdditionalDocument;
+        }
+
+        private void OpenDocument(DocumentId documentId, bool isAdditionalDocument)
+        {
+            if (isAdditionalDocument)
+            {
+                _previewWorkspace.OpenAdditionalDocument(documentId);
```

Comment:
workspace should have enough information to see whether given documentId is additional document or not.